core/src/apps/bitcoin/verification.py

```diff
@@ -0,0 +1,119 @@
+                    raise wire.DataError("Invalid public key hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys = [public_key]
+                self.signatures = [(signature, hash_type)]
+            elif len(script_sig) == 35:  # P2WSH nested in BIP16 P2SH
+                script, self.signatures = parse_witness_p2wsh(witness)
+                script_hash = sha256(script).digest()
+                if input_script_p2wsh_in_p2sh(script_hash) != script_sig:
+                    raise wire.DataError("Invalid script hash")
+                script_hash = coin.script_hash(script_sig[1:])
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = parse_output_script_multisig(script)
+            else:
+                raise wire.DataError("Unsupported signature script")
+        else:
+            if len(script_pubkey) == 25:  # P2PKH
+                public_key, signature, hash_type = parse_input_script_p2pkh(script_sig)
+                pubkey_hash = ecdsa_hash_pubkey(public_key, coin)
+                if output_script_p2pkh(pubkey_hash) != script_pubkey:
+                    raise wire.DataError("Invalid public key hash")
+                self.public_keys = [public_key]
+                self.signatures = [(signature, hash_type)]
+            elif len(script_pubkey) == 23:  # P2SH
+                script, self.signatures = parse_input_script_multisig(script_sig)
+                script_hash = coin.script_hash(script)
+                if output_script_p2sh(script_hash) != script_pubkey:
+                    raise wire.DataError("Invalid script hash")
+                self.public_keys, self.threshold = parse_output_script_multisig(script)
+            else:
+                raise wire.DataError("Unsupported signature script")
+
+        if self.threshold != len(self.signatures):
+            raise wire.DataError("Invalid signature")
+
+    def ensure_hash_type(self, hash_type: int) -> None:
+        if any(h != hash_type for _, h in self.signatures):
+            raise wire.DataError("Unsupported sighash type")
+
+    def verify(self, digest: bytes) -> None:
+        # It is up to the caller to ensure that the digest is appropriate for
+        # the hash type used by the signatures. If a signature is using a
+        # different hash type than expected, then verification will fail. To
+        # return the proper error message, the caller can optionally check the
+        # hash type by using ensure_hash_type() before calling verify.
+        try:
+            i = 0
+            for der_signature, _ in self.signatures:
+                signature = decode_der_signature(der_signature)
```

Comment:
I made a small optimization here for multisig to avoid repeated decoding: a2c47edbdcc3b3e7284449525d8378de6818e218, b6c8b8b88b134d6d52680955cc900d63ea34282c.